packages/itwin/tree-widget/src/tree-widget-react/components/trees/common/components/TreeNodeVisibilityButton.tsx

```diff
@@ -0,0 +1,83 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import "./TreeNodeVisibilityButton.css";
+import cx from "classnames";
+import { IconButton, Tooltip } from "@itwin/itwinui-react/bricks";
+import { isPresentationHierarchyNode } from "@itwin/presentation-hierarchies-react";
+
+import type { PresentationHierarchyNode, PresentationTreeNode } from "@itwin/presentation-hierarchies-react";
+import type { ComponentPropsWithoutRef } from "react";
+
+const visibilityShowIcon = new URL("@itwin/itwinui-icons/visibility-show.svg", import.meta.url).href;
+const visibilityHideIcon = new URL("@itwin/itwinui-icons/visibility-hide.svg", import.meta.url).href;
+const visibilityPartialIcon = new URL("@itwin/itwinui-icons/state-inherited-dot.svg", import.meta.url).href;
+
+/**
+ * Data structure that describes tree node checkbox state.
+ * @beta
+ */
+interface TreeItemVisibilityButtonState {
+  state: "visible" | "partial" | "hidden";
+  isDisabled?: boolean;
+  tooltip?: string;
+}
+
+/** @beta */
+export interface TreeItemVisibilityButtonProps {
+  /** Callback that should be invoked when checkbox is clicked. */
+  onVisibilityButtonClick: (node: PresentationHierarchyNode, state: TreeItemVisibilityButtonState["state"]) => void;
+  /** Callback that should be used to determine current checkbox state. */
+  getVisibilityButtonState: (node: PresentationHierarchyNode) => TreeItemVisibilityButtonState;
+}
+
+/** @internal */
+type TreeNodeCheckboxProps = TreeItemVisibilityButtonProps &
+  Omit<ComponentPropsWithoutRef<typeof IconButton>, "onClick" | "aria-disabled" | "title" | "label" | "icon">;
+
+/** @internal */
+export function TreeItemVisibilityButton({
+  node,
+  onVisibilityButtonClick,
+  getVisibilityButtonState,
+  ...props
+}: TreeNodeCheckboxProps & { node: PresentationTreeNode }) {
+  if ("type" in node || !isPresentationHierarchyNode(node)) {
+    return null;
+  }
+  const checkboxState = getVisibilityButtonState(node);
+
+  const getIcon = () => {
+    switch (checkboxState.state) {
+      case "visible":
+        return visibilityShowIcon;
+      case "hidden":
+        return visibilityHideIcon;
+      case "partial":
+        return visibilityPartialIcon;
+    }
+  };
+  return (
+    <TooltipWrapper content={checkboxState.tooltip}>
```

Comment:
Tooltip should not be necessary if you are using `IconButton` with label

Reply:
This is a itwinUI issue, label is required and generates a tooltip, our tooltip loads a bit later which causes it to be undefined for a bit, which means we can't use it as a label. We probably want a label and a separate tooltip which.

Reply:
changed